BUGFIX: Fusion's FilePatternResolver windows compatible

mhsdesign opened the pull request:
resolves #4358 in combination with #4374


The FilePatternResolver is responsible to handle glob patterns in includes in fusion.
For more in depth information see the test: https://github.com/neos/neos-development-collection/blob/a461dbc8f976588b30f34a5ed34f1eeebf5a0cd2/Neos.Fusion/Tests/Unit/Core/Parser/ParserIncludeTest.php#L156-L167

<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
BUGFIX: Fusion's FilePatternResolver windows compatible

resolves #4358

dlubitz commented on Neos.Fusion/Classes/Core/ObjectTreeParser/FilePatternResolver.php:
```diff
@@ -56,13 +57,14 @@ class FilePatternResolver
      * @throws Fusion\Exception
      */
-    public static function resolveFilesByPattern(string $filePattern, ?string $filePathForRelativeResolves = null, string $defaultFileEndForUnspecificGlobbing = '.fusion'): array
+    public static function resolveFilesByPattern(string $filePattern, ?string $filePathForRelativeResolves, string $defaultFileEndForUnspecificGlobbing): array
```
Why do you remove the defaults here? This would be a breaking change.

mhsdesign replied:
No its internal (was a protected method before Neos 8) ;) And my previous ME liked defaults, i hate em now ^^

dlubitz commented on Neos.Fusion/Classes/Core/ObjectTreeParser/FilePatternResolver.php:
```diff
@@ -56,13 +57,14 @@ class FilePatternResolver
             throw new Fusion\Exception("cannot resolve empty pattern: '$filePattern'", 1636144288);
         }
-        if ($filePattern[0] === '/') {
+        $isAbsoluteWindowsPath = str_contains($filePattern, ':') && preg_match('`^[a-zA-Z]:/[^/]`', $filePattern) === 1;
+        if ($filePattern[0] === '/' || $isAbsoluteWindowsPath) {
```
Maybe a new method `isAbsolutePath` in `Files` could cover this logic?

mhsdesign replied:
was thinking, the same but it lives in flow https://github.com/neos/flow-development-collection/blob/8.3/Neos.Utility.Files/Classes/Files.php and that would be cruel ^^

kdambekalns replied:
> and that would be cruel ^^

What would be cruel? And why?

mhsdesign replied:
because we dont specify the exact flow patch wed need 

https://github.com/neos/neos/blob/851c073c49c0d9e85f7081cc939c159401286343/composer.json#L11

https://github.com/neos/fusion/blob/5b45a833a1d8e900bbe04d18b31aed37c98aa7e6/composer.json#L10

so updating fusion alone would work (by composer constraints) but would fail at runtime because flow isnt at the latest patch ... then again who does this ...

kdambekalns replied:
Then you could require Flow in a matching minimum version…

dlubitz reviewed (commented):
Looks pretty reasonable. I left some comments to discuss.

mhsdesign commented on Neos.Fusion/Classes/Core/ObjectTreeParser/FilePatternResolver.php:
```diff
@@ -56,13 +57,14 @@ class FilePatternResolver
             throw new Fusion\Exception("cannot resolve empty pattern: '$filePattern'", 1636144288);
         }
-        if ($filePattern[0] === '/') {
+        $isAbsoluteWindowsPath = str_contains($filePattern, ':') && preg_match('`^[a-zA-Z]:/[^/]`', $filePattern) === 1;
```
> str_contains($filePattern, ':') &&

premature performance optimization is the root of all evil?

... anyways ...

dlubitz replied:
😈

mhsdesign replied:
well regex is slow, and most included paths are relative, so i think its a fine assumption of mine to speed up the happy path ^^

mhsdesign commented:
@dlubitz did .. did you actually test it :omg: or just fine by reading?

mhsdesign commented:
otherwise i might find someplace an windows mashine ^^

dlubitz commented:
No sorry, just by reading. Even if I have a window machine, I have no PHP running there. #wsl2

mhsdesign commented:
<img width="403" alt="image" src="https://github.com/neos/neos-development-collection/assets/85400359/c1553352-76f7-41fa-9e8b-7f5732860e9a">

crydotsnake commented:
Maybe @patricekaufmann could test this. He is using Windows too afaik.

kdambekalns reviewed (commented):
Generally like this. Cannot test this. But I wonder if there is really no way to unit test this?

patricekaufmann commented:
> Maybe @patricekaufmann could test this. He is using Windows too afaik.

I am using Windows but in 2019 I started using WSL2, so there is no real way for me to test this as I would have to look up how to actually setup a proper development environment on windows.

mhsdesign committed:
TASK: Backport fix from 6104891f44394f191563ced7019bd42265a98cb0

mhsdesign commented:
after spending the evening getting it run on windows 😅 , i could confirm that this fixes the issue partially in combination with https://github.com/neos/neos-development-collection/pull/4374

kdambekalns reviewed (approved):
If you sya it works… Looks good, even though I'd love to make use of the `Files` utility class in Flow. 🤷‍♂️

mhsdesign committed:
BUGFIX: `nodetypes://` normalize the path to unix

`nodetypes://Foo.Bar/**/*.fusion`

was resolved to the path:

`nodetypes://Foo.Bar\Content\File.fusion`

this is invalid to the nodeTypes stream wrapper, as we split on unix slash `/`. Backslashes are invalid syntax.

mhsdesign committed:
BUGFIX: FilePatternResolver dont allow non-unix style paths
